pallets/token-factory/src/lib.rs

```diff
@@ -0,0 +1,428 @@
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use ethereum_types::BigEndianHash;
+	use fp_evm::ExecutionInfo;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use pallet_evm::{AddressMapping, ExitReason, Runner};
+	use parity_scale_codec::{Decode, Encode, FullCodec};
+	use rustc_hex::FromHex;
+	use sp_core::{H160, H256, U256};
+	use sp_runtime::{
+		traits::{AtLeast32BitUnsigned, Convert, Zero},
+		DispatchError, SaturatedConversion,
+	};
+	use sp_std::{fmt::Debug, vec::Vec};
+
+	/// ERC20PresetMinterBurner contract bytecode
+	const CONTRACT_BYTECODE: &str = include_str!("../contract/bytecode.txt");
+
+	/// The ERC token factory pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_evm::Config + pallet_sudo::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Balances type
+		type Balance: Parameter
+			+ Member
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ Into<u128>;
+		/// Token identifier
+		type TokenId: Clone + Copy + FullCodec + Debug + PartialEq + Ord;
+		/// Convert from AccountId to H160, is identity map for Moonbeam
+		type AccountToH160: Convert<Self::AccountId, H160>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// TokenID is already claimed
+		IdClaimed,
+		/// TokenID does not exist so cannot interact with it
+		IdNotClaimed,
+	}
+
+	#[derive(PartialEq, Clone, Copy, Encode, Decode, sp_runtime::RuntimeDebug)]
+	pub enum EvmCall {
+		Register,
+		Mint,
+		Burn,
+		TotalIssuance,
+		BalanceOf,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Token register success [token_id, contract_address]
+		Registered(T::TokenId, H160),
+		/// Mint token success. [token_id, who, amount]
+		Minted(T::TokenId, T::AccountId, T::Balance),
+		/// Burn token success. [token_id, who, amount]
+		Burned(T::TokenId, T::AccountId, T::Balance),
+		/// Destroy all tokens success. [token_id, amount]
+		DestroyedAll(T::TokenId, T::Balance),
+		/// Call failed with exit reason [call, reason]
+		EvmCallFailed(EvmCall, ExitReason),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::storage]
+	#[pallet::getter(fn nonce)]
+	pub type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn tokens)]
+	pub type Tokens<T: Config> = StorageValue<_, Vec<T::TokenId>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn contract_address)]
+	pub type ContractAddress<T: Config> =
+		StorageMap<_, Twox64Concat, T::TokenId, H160, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig {
+		pub nonce: U256,
```

Comment:
add tokens: Vec<T::TokenId> with default empty vec

Reply:
I tried this, but I couldn't deploy to the EVM in genesis build, something to do with nested transactions in storage unimplemented so registration will need to be manual